app/coffee/RedisManager.coffee

```diff
@@ -287,6 +287,34 @@ module.exports = RedisManager =
 	getDocIdsInProject: (project_id, callback = (error, doc_ids) ->) ->
 		rclient.smembers keys.docsInProject(project_id: project_id), callback
 
+	getDocTimestamps: (doc_ids, callback = (error, result) ->) ->
+		# get lastupdatedat timestamps for an array of doc_ids
+		multi = rclient.multi()
+		for doc_id in doc_ids
+			multi.get keys.lastUpdatedAt(doc_id: doc_id)
+		multi.exec callback
+
+	queueFlushAndDeleteProject: (project_id, callback) ->
+		# store the project id in a sorted set ordered by time
+		rclient.zadd keys.flushAndDeleteQueue(), Date.now(), project_id, callback
+
+	getNextProjectToFlushAndDelete: (cutoffTime, callback = (error, key, timestamp)->) ->
+		# find the oldest queued flush that is before the cutoff time
+		rclient.zrangebyscore keys.flushAndDeleteQueue(), 0, cutoffTime, "WITHSCORES", "LIMIT", 0, 1, (err, reply) ->
+			return callback(err) if err?
+			return callback() if !reply?.length # return if no projects ready to be processed
+			# pop the oldest entry (get and remove in a multi)
+			multi = rclient.multi()
+			multi.zrange keys.flushAndDeleteQueue(), 0, 0, "WITHSCORES"
```

Comment:
Would [ZPOPMIN](https://redis.io/commands/zpopmin) work here?

Reply:
Unfortunately it is Redis 5 only (as I discovered after starting out using it).